Port everything to TypeScript

brownjava opened the pull request:
This is my first crack at porting everything over to TypeScript.  What's done:

- The project compiles with tsc
- Tests all run

What's not tested:

- Using it from other projects, specifically those that don't use imports.

I wanted to get this change out now in an early state because it's dragged on more than I would have liked, and I want to get some input on how folks feel about TypeScript.  Namely is the overhead of typing worth the wins.

tomborger commented:
I like it. It doesn't seem unreasonably verbose, and I love the clearly defined interfaces and abstract classes. The import/require interop is nice too. Even if there's an adjustment period (there definitely will be for me) I think it's good for us to get on this train sooner rather than later.

brownjava commented:
I've added tslint and cleaned things up quite a bit.  At this point I think we are ready to merge this in, bump the minor number (this is too big for just bumping the patch number), and release.
